File: src-tauri/plugins/cosmos/android/src/main/java/lat/cosmospay/plugin/cosmos/ShareSheet.kt
package lat.cosmospay.plugin.cosmos

import android.app.Activity
import android.content.Intent

/**
 * The Android share sheet.
 *
 * Shares a crate with the secure store and nothing else — see the header on
 * `src-tauri/plugins/cosmos/src/lib.rs`. It exists because an Android WebView does not
 * implement `navigator.share`, so the one platform with a real share sheet was the one
 * that silently fell back to copying.
 */
internal object ShareSheet {

    /**
     * Raise the chooser.
     *
     * `createChooser` rather than the bare `ACTION_SEND`: the bare intent throws
     * `ActivityNotFoundException` on a device with nothing that accepts `text/plain`, and
     * on a device with exactly one such app it silently skips the choice the user was
     * offered. The chooser is itself an activity, so it always resolves.
     *
     * The whole call is wrapped by the caller: what reaches the wallet is `false`, which
     * makes it copy to the clipboard instead — see `src/lib/share.ts`.
     */
    fun share(activity: Activity, text: String, title: String?) {
        val send = Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            putExtra(Intent.EXTRA_TEXT, text)
            // A subject is what an email client fills its subject line with. Absent for a
            // messenger, which is the common case, so it stays optional rather than being
            // defaulted to something the wallet made up.
            if (title != null) putExtra(Intent.EXTRA_SUBJECT, title)
        }
        activity.startActivity(Intent.createChooser(send, title))
    }
}

File: src-tauri/plugins/cosmos/ios/Sources/CosmosPlugin/ShareSheet.swift
import UIKit

/// The iOS share sheet.
///
/// Shares a package with the secure store and nothing else — see the header on
/// `src-tauri/plugins/cosmos/src/lib.rs`. It exists because a WKWebView does not implement
/// `navigator.share`, so the one platform with a real share sheet was the one that
/// silently fell back to copying.
enum ShareSheet {

    /// Present `UIActivityViewController` over the app.
    ///
    /// Main thread only — UIKit presentation is not thread-safe and the caller reaches this
    /// from a background queue.
    ///
    /// The completion fires once the sheet closes, but the wallet is told the share
    /// SUCCEEDED as soon as the sheet appears. That is on purpose: `src/lib/share.ts` reads
    /// `false` as "no sheet exists, copy instead", and a user who opens the sheet and then
    /// changes their mind must not have their clipboard overwritten behind their back.
    static func present(from controller: UIViewController, text: String, title: String?) throws {
        let activity = UIActivityViewController(activityItems: [text], applicationActivities: nil)

        // An iPad presents this as a popover and CRASHES if it has no anchor. There is no
        // originating view here — the request came from the web layer — so it is anchored
        // to the middle of the presenting controller, with an empty rect so no arrow is
        // drawn pointing at nothing in particular.
        if let popover = activity.popoverPresentationController {
            popover.sourceView = controller.view
            popover.sourceRect = CGRect(
                x: controller.view.bounds.midX,
                y: controller.view.bounds.midY,
                width: 0,
                height: 0
            )
            popover.permittedArrowDirections = []
        }

        if let title {
            activity.setValue(title, forKey: "subject")
        }

        controller.present(activity, animated: true)
    }
}
